pydepsi/io.py

```diff
@@ -0,0 +1,351 @@
+"""io methods."""
+
+import re
+from datetime import datetime
+
+import numpy as np
+
+# Define constants
+SC_N_PATTERN = r"\s+([-+]?\d+(?:\.\d+)?(?:[eE][-+]?\d+)?)"
+SPPED_OF_LIGHT = 299792458.0
+
+
+def read_metadata(resfile, mode="raw", **kwargs):
+    """Read metadata from a DORIS v5 resfile.
+
+    Modified from the original functions in:
+    https://github.com/Pbaz98/Caroline-Radar-Coding-Toolbox/blob/main/gecoris/dorisUtils.py.
+    """
+    # check crop_flag
+    if mode == "coreg" and "crop" in kwargs:
+        crop_flag = kwargs["crop"]
+    else:
+        crop_flag = 0
+
+    # Open the file
+    with open(resfile) as file:
+        content = file.read()
+
+    # +++   - Satellite ID
+    pattern = r"Product type specifier:\s+(.*?(?=\n))"
+    match = re.search(pattern, content)
+    match.group(1).upper()
+
+    # ++++1 - Geometry [DESCENDING or ASCENDING]
+    pattern = r"PASS:\s+(.*?(?=\n))"
+    match = re.search(pattern, content)
+    geometry = match.group(1).upper()
+
+    # ++++ 2 - Acquisition Date [dictionary with __datetime__ and str, in the format 'yyyy-mm-dd hh:mm:ss']
+    pattern = r"First_pixel_azimuth_time \(UTC\):\s+(\d+-\w+-\d+\s+)(\d+:\d+:\d+)"
+    match = re.search(pattern, content)
+
+    # --- extract the datetime string
+    datetime_toconvert = match.group(1) + match.group(2)
+    # Parse the original datetime string
+    acq_date = datetime.strptime(datetime_toconvert, "%Y-%b-%d %H:%M:%S")
+    acq_date.strftime("%Y%m%d")
+
+    # ++++ 3 - azimuth0time
+
+    # convert from time format to seconds of the day
+    match.group(2)
+    pattern = r"(\d+):(\d+):(\d+.\d+)"
+    match = re.search(pattern, content)
+    azimuth0time = int(match.group(1)) * 3600 + int(match.group(2)) * 60 + float(match.group(3))
+
+    # ++++ 4 - range0time
+    pattern = r"Range_time_to_first_pixel \(2way\) \(ms\):" + SC_N_PATTERN
+    match = re.search(pattern, content)
+    range0time = float(match.group(1)) * 1e-3
+
+    # ++++ 5 - prf
+    pattern = r"Pulse_Repetition_Frequency \(computed, Hz\):" + SC_N_PATTERN
+    match = re.search(pattern, content)
```

Comment:
There is a huge number of the regular expression searches here and in the following. I'd suggest moving this to a separate module and making a wrapper which calls them by name, allowing the correct implementation to be retrieved. This also makes one central place for keeping track of all these things and easily supports adding additional methods as needed.
Happy to discuss @rogerkuou

Reply:
Hi @meiertgrootes, I like this idea! I added your suggestion as an item in issue #14. My thought is to build a static dict, mapping the attribute name and searching pattern. Let me know what you think.